controllers/cache_handler.go

```diff
@@ -0,0 +1,497 @@
+	}
+}
+
+// OnChangeHandleNotifierSink Trigger Notifier Sink on changes in the targets
+func (h DisruptionTargetWatcherHandler) OnChangeHandleNotifierSink(oldPod, newPod *corev1.Pod, oldNode, newNode *corev1.Node, okOldPod, okNewPod, okOldNode, okNewNode bool) {
+	var objectToNotify runtime.Object
+
+	eventsToSend := make(map[string]bool)
+	name := ""
+
+	switch {
+	case okNewPod && okOldPod:
+		objectToNotify, name = newPod, newPod.Name
+
+		disruptionEvents, err := h.getEventsFromCurrentDisruption("Pod", newPod.ObjectMeta, h.disruption.CreationTimestamp.Time)
+		if err != nil {
+			h.reconciler.log.Warnf("couldn't get the list of events from the target. Might not be able to notify on error changes: %s", err.Error())
+		}
+
+		// we detect and compute the error / warning events, status changes, conditions of the updated pod
+		eventsToSend = h.buildPodEventsToSend(*oldPod, *newPod, disruptionEvents)
+	case okNewNode && okOldNode:
+		objectToNotify, name = newNode, newNode.Name
+
+		disruptionEvents, err := h.getEventsFromCurrentDisruption("Node", newNode.ObjectMeta, h.disruption.CreationTimestamp.Time)
+		if err != nil {
+			h.reconciler.log.Warnf("couldn't get the list of events from the target. Might not be able to notify on error changes: %s", err.Error())
+		}
+
+		// we detect and compute the error / warning events, status changes, conditions of the updated node
+		eventsToSend = h.buildNodeEventsToSend(*oldNode, *newNode, disruptionEvents)
+	default:
+		h.reconciler.log.Debugw("target observer couldn't detect what type of changes happened on the targets")
+	}
+
+	// Send events to notifier / to disruption
+	lastEvents, _ := h.getEventsFromCurrentDisruption("Disruption", h.disruption.ObjectMeta, h.disruption.CreationTimestamp.Time)
+
+	for eventReason, toSend := range eventsToSend {
+		if !toSend {
+			continue
+		}
+
+		eventType := corev1.EventTypeWarning
+		if eventReason == chaosv1beta1.EventNodeRecoveredState || eventReason == chaosv1beta1.EventPodRecoveredState {
+			eventType = corev1.EventTypeNormal
+		}
+
+		// Send to updated target
+		h.reconciler.Recorder.Event(objectToNotify, eventType, eventReason, fmt.Sprintf(chaosv1beta1.Events[eventReason].OnTargetTemplateMessage, h.disruption.Name))
+
+		// Send to disruption, broadcast to notifiers
+		for _, event := range lastEvents {
+			if event.Type == eventReason {
+				h.reconciler.Recorder.Event(h.disruption, eventType, eventReason, fmt.Sprintf(chaosv1beta1.Events[eventReason].OnDisruptionTemplateMessage, name))
+
+				return
+			}
+		}
+
+		h.reconciler.Recorder.Event(h.disruption, eventType, eventReason, chaosv1beta1.Events[eventReason].OnDisruptionTemplateAggMessage)
+	}
+}
+
+func (h DisruptionTargetWatcherHandler) getEventsFromCurrentDisruption(kind string, objectMeta v1.ObjectMeta, disruptionStateTime time.Time) ([]corev1.Event, error) {
```

Comment:
We are ok on the tradeoff that if a disruption event happens while the controller is down, this event will be skipped (which is to me better than renotifying about something that happened in the past). Still need to be confirmed by tests.